### PR TITLE
support for ESP32-S2

### PR DESCRIPTION
**Pull request by mmoskal**
General:
* allow `net.ControllerSocket` to be event-driven

VM work:
* support for suspend/resume of fibers
* fix for verification error 1063

ESP32 / ESP32-S2 work:
* add wifi (SSL mostly) driver
* add simple Pin driver
* add USB driver (serial only); add HF2 over serial
* reset to booloader on GPIO0 (button) for S2
* settings module (using NVS storage)

### CODE REVIEW

File: libs/base/pxtbase.h [hunk 1]
```diff
@@ -1047,14 +1050,15 @@ template <size_t N> constexpr size_t _boxedStringLen(char const (&)[N]) {
...
+    const BoxedStringLayout<_boxedStringLen(val)> name[1] = {                                      \
         {&pxt::string_inline_ascii_vt, _boxedStringLen(val) - 1, val}};
 
 // bigger value - less memory, but slower
 // 16/20 keeps s.length and s.charCodeAt(i) at about 200 cycles (for actual unicode strings),
 // which is similar to amortized allocation time
 #define PXT_STRING_SKIP_INCR 16 // needs to be power of 2; needs to be kept in sync with compiler
-#define PXT_STRING_MIN_SKIP 20  // min. size of string to use skip list; static code has its own limit
+#define PXT_STRING_MIN_SKIP                                                                        \
+    20 // min. size of string to use skip list; static code has its own limit
```
Comment (pelikhan) [inline libs/base/pxtbase.h:1061]: new C code formatter?

**Review by pelikhan — APPROVED** [review]

File: libs/wifi---esp32/ca-bundle.crt [hunk 2]
```diff
@@ -0,0 +1,3191 @@
+##
```
Comment (pelikhan) [inline libs/wifi---esp32/ca-bundle.crt:1]: do you have more docs about this file, like how to generate it and so forth?

Comment (mmoskal) [reply libs/wifi---esp32/ca-bundle.crt]: added instructions

File: libs/base/pxtbase.h (hunk 1 above)
Comment (mmoskal) [reply libs/base/pxtbase.h:1061]: no; I suspect some manual changes without reformatting in the past...

**Pull request merged**